Redesign getindex and fix stack/melt StackOverflow error

bkamins opened the pull request:
Currently `stack(df, :a, [])` produces StackOverflow.
With this PR we check that length of `measure_vars` and `id_vars` is at least 1.

bkamins commented:
I had to additionally fix `getindex` implementation here (which could lead to other errors also):
https://github.com/JuliaData/DataFrames.jl/pull/1305/files#diff-084f63e064c680ae660e2676e15d3d4eR116

bkamins commented:
After looking at the implementation I recommend to allow empty sets measures or ids. So now the only change that I make is fixing `getindex`.

nalimilan commented:
Can you explain what's the use of empty measures of IDs? Can you check what other implementations (like R, tidyr and Pandas) do in that case?

bkamins commented:
I do not know the intention of the original code, but how it works is the following:
* empty id: simply no id column is created (this is logical)
* empty measures: 0 rows are created (this is logical)

I have checked `melt` from `reshape2` and it allows both to be empty. For empty id it works the same as above. For empty measures it produces an original vector of id only (this seems wrong to me). For empty id and empty measures it produces empty DataFrame.

CC @tshort (as I understand you have written the original code so maybe you have some comments on it)

bkamins commented:
This PR is in interaction with #1306. I propose that I wait till the other one is merged, then I rebase this one and will fix errors in tests.

bkamins commented:
rebased. A review now would be welcome 😄.

nalimilan commented on src/other/index.jl:
```diff
@@ -113,8 +113,13 @@ Base.getindex(x::AbstractIndex, idx::Real) = Int(idx)
-Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T >: Missing} =
-    getindex(x, collect(skipmissing(idx)))
+function Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T >: Missing}
+    idx2 = collect(skipmissing(idx)) # idx can be Any and we have to handle it
```
Actually I think there would be a cleaner and more efficient way to do this: 
```julia
Int[i isa Real ? i : x[i] for i in skipmissing(idx)]
```

bkamins replied:
After thinking about it I propose an implementation which disallows mixing `Real` and `Symbol` indexing, so that something like this `df[[1, :x2, missing]]` will throw an error (in the old implementation it would go through).

cjprybol commented:
I don't think this is an issue of having empty measures/ids so much as it is an issue of type restriction. `stack(df, :a, [])` is passing an array of type `Any`, which is causing the `Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T >: Missing}` function to call itself over and over. I think it makes more sense to restrict the types and throw a `MethodError` when anyone tries to pass a Vector of `Any`, rather than try and handle whatever values might end up in an `Any[]`. We could also restrict the type signature of `stack`, but as an alternative, see https://github.com/JuliaData/DataFrames.jl/pull/1311 and let me know what you think

nalimilan commented:
I wish we could reject `Array{Any}`, but it's supported in Base when indexing vectors, and not accepting them makes the behavior more fragile, especially since currently `typejoin(T, Missing) === Any`.

cjprybol commented:
Ah, nevermind then. Carry on

bkamins commented on test/data.jl:
```diff
@@ -143,6 +143,20 @@ module TestData
                    d = Array{Union{Float64, Missing}}(randn(12)),
                    e = Array{Union{String, Missing}}(map(string, 'a':'l')))
 
+    # test empty measures or ids
```
moved

bkamins commented:
Following this comment I have made `DataFrame` and `Index` accept arrays of `Any` (for `DataFrames` also for row indexing).

The redesign actually makes minimal signatures of `getindex` for `DataFrame` - the logic to test if input is allowed is passed to downstream `getindex` functions.

For `Index` implementation of `getindex` I make sure that we do not allow mixing numbers and symbols in the vector (I guess this is what is a sensible behavior).

Additionally (maybe for another PR) `getindex` for `DataFrame`:
* throws error if `missing` is encountered in row-index;
* removes `missing`s if they are encountered in column-index.

This is inconsistent. My intuition would be not to allow `missing` in column-indices and throw error when they are encountered (this is how row-index works and how Base works), but maybe there are some good reasons to allow it?

nalimilan commented on src/other/index.jl:
```diff
@@ -113,11 +113,20 @@ Base.getindex(x::AbstractIndex, idx::Real) = Int(idx)
 Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
+function Base.getindex(x::AbstractIndex, idx::AbstractVector)
+    idxs = skipmissing(idx)
+    try # allow either vector of numbers
```
I don't really like that approach. It would be simpler to do `[convert(Union{Int, Symbol}, i) for i in idxs]`. If we really don't want to allow mixing integers and symbols, we can always check the type of the result at the end.

bkamins replied:
fixed

bkamins commented:
@nalimilan - do you feel that we should allow `missing` when indexing columns? Why was it introduced?

nalimilan commented:
> @nalimilan - do you feel that we should allow missing when indexing columns? Why was it introduced?

No, I don't think that makes sense. I don't even know what it does. We don't allow for missing values anywhere else in indexing operations AFAIK.

bkamins commented:
I will fix the tests and a separate PR for removing support for `missing`.

bkamins commented:
If this is merged (and #1326 and #1325) I will remove support for indexing with `missing`.
There will probably be some rebasing needed between those PRs, so they have to be analyzed sequentially.

bkamins commented:
rebased and added deprecation for using `missing` in column indexing (actually the required rewrite of `getindex` made it natural to include it here).

nalimilan commented on src/other/index.jl:
```diff
@@ -110,17 +110,31 @@ end
+    length(idxs) == 0 && return Int[] # special case of empty idxs
+    # below is not very efficient, but has to be done as in Bool case we replaced missing by false
+    # after deprecation period when Missing is disallowed this will not be a problem
+    idxs[1] isa Bool && return getindex(x, Bool[ismissing(v) ? false : v for v in idx])
```
I think we'd better handle logical indexing only in the `AbstractVector{Bool}`/`AbstractVector{<: Union{Bool, Missing}}` method. That's what is done for `Array` in Base. No need to make this more complex.

bkamins replied:
Agreed - in 0.7 indexing with `true` is deprecated (in 0.6 it was not). I will remove this support.

A minor question is why do we convert any `Real` to `Int` while Base allows only `Integer`?

nalimilan replied:
I guess that's historical. We could deprecate it.

nalimilan commented on src/other/index.jl:
```diff
@@ -110,17 +110,31 @@ end
+    idxs[1] isa Real && return Vector{Int}(idxs)
+    idxs[1] isa Symbol && return [x.lookup[i] for i in idxs]
+    throw(ArgumentError("idx[1] has type $(typeof(idx[1]));"*
+                        "getindex supports only indexing by integers, booleans or symbols"))
```
I'd rather say "data frames only support indexing columns with integers, symbols or boolean vectors".

bkamins replied:
fixed

bkamins commented:
I have also deprecated `df[true]` as it is deprecated in 0.7 Base.

bkamins commented:
I have added the deprecations. Now `getindex` for `Index`:
* gives guarantee that it returns either `Int` or `Vector{Int}` (in the old implementation this invariant was not satisfied by `AbstractRange` version of `getindex`).
* does not check if passed indices are inbounds.

Actually I am tempted to recommend the following target design:
```
Base.getindex(x::AbstractIndex, idx::Symbol) = x.lookup[idx]
Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
Base.getindex(x::AbstractIndex, idx) = (1:length(x))[idx]
```
it would be cleanest and delegated handling of `idx` other than `Symbol` to array indexing in Base (so that we do not have to think if we are consistent with it or not).
The consequences are:
* we would get checking bounds of `idx`; this is a bit of overhead but I assume it would be negligible (if we were very sensitive to performance `Base.getindex(x::AbstractIndex, idx::AbstractVector{Int}) = idx` could be defined and it would not check bounds or do any copying);
* when indexing with vector of `Sybmol` user would have to make sure its type is `Vector{Symbol}`, but I think it should not be a big problem.

nalimilan commented:
I'm trying to move to a design based on `NamedTuple`, we could refactor the indexing code in that occasion. Ideally we wouldn't need the `Index` type at all (but I'm not sure that's possible).

nalimilan commented on src/other/index.jl:
```diff
@@ -108,19 +108,49 @@ function Base.insert!(x::Index, idx::Integer, nm::Symbol)
+Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
+
+function Base.getindex(x::AbstractIndex, idx::Real)
+    idx isa Bool && Base.depwarn("Indexing with Bool values is deprecated", :getindex)
```
Better implement these deprecations via specialized methods in deprecated.jl, so that the code here doesn't need to be changed too much to remove deprecations.

Also, "Indexing with Bool values is deprecated" could be misleading: should specify "except for `Array{Bool}`".

bkamins replied:
OK. The message is taken from Base but I will add the additional comment.

nalimilan commented on src/other/index.jl:
```diff
@@ -108,19 +108,49 @@ function Base.insert!(x::Index, idx::Integer, nm::Symbol)
-Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T <: Real} = convert(Vector{Int}, idx)
-Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
+
+Base.getindex(x::AbstractIndex, idx::AbstractRange) = getindex(x, collect(idx))
```
The old method was more efficient, why call `getindex`?

bkamins replied:
because `[idx;]` may produce something else that `Vector{Int}` and the contract of `getindex` is that it returns either `Int` or `Vector{Int}`.

nalimilan replied:
But not why just `convert(Vector{Int}, idx)`? More radically, why not change the contract and decide it can return any `AbstractVector{Int}`? Doesn't sound very useful to convert to a vector.

bkamins replied:
It is not a `Vector` part that is a problem, but `Int` part.

If I simply convert then `1.0:3.0` or `true:true` will silently pass though as valid index ranges and the should not (at least based on what is consistent with Base). Currently it is converted to `Vector` so the second round of dispatching does it work to check if indexing is valid.

Exactly because of such problems  I think that in general it is better to delegate handling of indexing to Base except what we have special in `DaraFrames` (i.e. indexing by `Symbol` or `AbstractVector{Symbol}`).

nalimilan replied:
Actually, wouldn't it be possible to have a method for `::AbstractRange{<:Integer}` which will just return its input, and one for `::AbstractRange` which will call `collect`?

bkamins replied:
`stack` in reshape.jl expects `Vector{Int}` explicitly now (and relies on the conversion made here). I can change it.
Also we would have to handle a special case of `:AbstractRange{Bool}` separately but it should not be a problem.

Should I go ahead with those changes?

nalimilan replied:
In general better require just `::AbstractVector{Integer}`, yes.

bkamins replied:
refactored. I continue to `collect` general ranges of `AbstractRange{<:Integer}` other than `AbstractRange{Int}` to cast them to `Int` (so that we have always underlying `Int` returned), but it can be changed.

bkamins commented:
This is exactly what I was thinking about. Pluses are that we get type inference for free.

Minuses are that:
* any change of column (adding, removing, changing type) forces us to rebuild whole named tuple;
* all functions working on `DataFrame` will have to recompile every time a new `DataFrame` is passed to them (unless you decide to hide `NamedTuple` inside a type and not expose it but then type inference will not work);

nalimilan commented on src/other/index.jl:
```diff
@@ -108,19 +108,49 @@ function Base.insert!(x::Index, idx::Integer, nm::Symbol)
+    end
+    length(idxs) == 0 && return Int[] # special case of empty idxs
+    # below is not very efficient, but has to be done as in Bool case we replaced missing by false
+    # after deprecation period when Missing is disallowed this will not be a problem
```
AFAICT we previously didn't support logical indexing via `AbstractVector{Any}`. So this deprecation could be implemented via a method for `AbstractArray{<:Union{Missing,Bool}}` in deprecated.jl to keep this one clean.

bkamins replied:
OK

nalimilan commented on src/other/index.jl:
```diff
@@ -108,19 +108,49 @@ function Base.insert!(x::Index, idx::Integer, nm::Symbol)
+        throw(ArgumentError("Indexing with Bool values is not allowed"))
+    end
+    idxs[1] isa Real && return getindex(x, Vector{Real}(idxs))
+    idxs[1] isa Symbol && return [x.lookup[i] for i in idxs]
```
Why not call `getindex` here too for simplicity and consistency?

bkamins replied:
OK, but this will introduce a bit of performance overhead.
I suppose it is not a huge problem as method with incorrect type of `Vector` should not be a typical scenario so I will change it.

nalimilan replied:
Are you sure? With inlining, shouldn't be code be identical?

bkamins replied:
The performance overhead is the following:
```
julia> x = Any[Symbol(s) for s in 'A':'z'];

julia> @benchmark convert(Vector{Symbol}, $x)
BenchmarkTools.Trial:
  memory estimate:  544 bytes
  allocs estimate:  1
  --------------
  minimum time:     3.382 μs (0.00% GC)
  median time:      3.616 μs (0.00% GC)
  mean time:        3.750 μs (0.43% GC)
  maximum time:     93.423 μs (89.08% GC)
  --------------
  samples:          10000
  evals/sample:     8
```
by using `convert` (or what I have written) - a new vector will be allocated (checked on 0.6.2 an 0.7) - and I need to convert to make the dispatch work.

nalimilan commented:
Actually I wanted to keep `DataFrame` type-unstable precisely to be able to add a remove columns, and have it wrap a type-stable object. But more on that once I have something mostly working.

bkamins commented:
I have refactored the code the way you have recommended, but the more I work with it the more I feel that we should use `getindex` from Base if possible (to be considered after the redesign you are working on).

nalimilan commented on src/other/index.jl:
```diff
@@ -108,19 +108,45 @@ function Base.insert!(x::Index, idx::Integer, nm::Symbol)
+
+function Base.getindex(x::AbstractIndex, idx::AbstractVector{<:Integer})
+    # TODO: this line should be changed to throw an error after deprecation
+    if any(v -> v isa Bool, idx)
```
You should be able to move this to deprecated.jl too with `::AbstractVector{Bool}`. Also better use `convert` below to avoid making a copy of `idx` when possible.

bkamins replied:
This may not go do deprecated.jl, because - at least as is - it will have to stay here and throw an error if `Bool` is in any entry of `idx`.

Given that above there is a method (line 114):
```
Base.getindex(x::AbstractIndex, idx::AbstractVector{Int}) = idx
```
when do you believe that in this method no conversion would be required?

nalimilan commented on src/deprecated.jl:
```diff
@@ -1285,3 +1285,18 @@ import Base: |>
+    Int(idx)
+end
+
+function Base.getindex(x::AbstractIndex, idx::AbstractVector{Union{Bool, Missing}})
```
Woops, actually I think we should allow `AbstractVector{Union{Bool, Missing}` but print a deprecation warning for missing values. That's kind of a special case which hasn't been considered in Base yet, but probably should for convenience.

bkamins replied:
What do you expect the functionality in Base will be then (so that I know what to implement).

But again - in general I think is better not to handle it at all but delegate to Base and treat all this code as transitory aiming in long run to write in the documentation something like:
> column indexing works like all indexing of vectors in Base except that also indexing by `Symbol` or `Vector{Symbol}` is allowed

I would even go further and maybe even target to say:
> column indexing works like all indexing of `NamedTuple` in Base

which can be done even if internally we do not switch to `NamedTuple` (but if we do then it will be given out-of-the-box). The difference is that `NamedTuple` does not support indexing by `Vector{Byte}` and `Vectors` do.

nalimilan replied:
> What do you expect the functionality in Base will be then (so that I know what to implement).

As I noted for now I think we should print a deprecation warning when missing values are found, replacing them with `false` as we do now to avoid breaking code. Then we could make them an error in the next major release.

It would be nice to be fully consistent with Base, but I'm not sure the indexing behavior of `NamedTuple` is completely settled yet. If we find inconsistencies we might want to discuss whether to change Base if we find it useful.

bkamins replied:
fixed

bkamins commented:
In summary I believe we should make `getindex` on columns of `DataFrame` make as close to Base as possible, as this is what users will expect. Otherwise there is a big mental burden to know what works where (while in particular row indexing now follows this pattern and uses rules from Base).

Just to show what we have now (I am using current release of DataFrames, i.e. 0.11.3):
```
julia> x = DataFrame(rand(3,3));

julia> x[[1.0,2],[1,2]]
ERROR: ArgumentError: invalid index: 1.0
# but:
julia> x[[1,2],[1.0,2]]
2×2 DataFrames.DataFrame
│ Row │ x1       │ x2       │
├─────┼──────────┼──────────┤
│ 1   │ 0.184084 │ 0.123543 │
│ 2   │ 0.855894 │ 0.45012  │
# but
julia> x[[1,2],1.0:2]
ERROR: ArgumentError: invalid index: 1.0
```
and there still a difference between `getindex` on `DataFrame` and on `index` (e.g. the last error is raised by `DataFrame` but not by `Index` `getindex`, and some internal functions in `DataFrames` use `getindex` of `Index` directly).

bkamins commented:
I guess we are close to Base functionality now. The summary how we work is:
* indexing by single `Symbol` returns a column number, guaranteed to be valid;
* indexing by single `Integer` returns a column number possibly invalid, except indexing by single `Bool` which is deprecated;
* indexing by single `Real` other than `Integer` converts it to `Int` and is deprecated;
* indexing by `Vector{Symbol}` returns a vector of column numbers, guaranteed to be valid;
* indexing by `Vector{<:Integer}` returns a vector of column numbers possibly invalid; it is deprecated to use `Bool` as entry in such a vector;
* indexing by `Vector{Bool}` or `Vector{Union{Bool, Missing}}` returns a vector of column numbers, guaranteed to be valid if length of the vector is equal to number of columns in `DataFrame`; using `missing` in such a vector changes it to `false` which is deprecated;
* indexing by `Vector` of any other type:
    * first drops `missing`s in this vector which is deprecated;
    * if resulting vector is empty then `Int[]` is returned
    * if resulting vector consists only of `Symbol` then indexing by `Vector{Symbol}` is applied;
    * if resulting vector consists only of `Real` then a vector of column numbers is returned indices may be invalid; elements of this vector that are not `Integer` or are `Bool` are converted to `Int`, which is deprecated;
    * otherwise an error is raised;
* indexing by `Range` falls back to indexing by `Vector` of an type inferred by `collect`;

I would leave the rest till the redesign of internals of `DataFrames` and `NamedTuples` settle in Base - but the changes should be small I guess (the major difference is that `NamedTuple` and `Tuple` do not support indexing by `Vector{Bool}` - I do not know if it is intentional or a missing functionality?).
